jenkins/_analysis.py

```diff
@@ -309,6 +340,102 @@ def get_classified_failures(build_data):
     return individual_failures
 
 
+def get_time_to_success(build_data):
+    """
+    Attempts to approximate how much time it would take each of the sub-builds
+    to reach an outcome conducive with merging a PR. This is an approximation
+    from the time that __main_multijob is originally pressed, until the
+    sub-build has been successfully concluded after retries.
+
+    As a baseline we take the duration of __main_multijob. An engineer should
+    not merge a branch in ideal conditions until __main_multijob is concluded.
+
+    If a job fails, we add to the time of __main_multijob the amount of time of
+    the next run of the same job. In the worst case a job fails and it was the
+    last job to run in __main_multijob. This approximates immediately hitting
+    retry on the failed job in that scenario. This is repeated until we get a
+    to a successful run of the job.
+
+    Note that if the most recent job is a failure, then this column will have
+    NaT in this column.
+
+    :param build_data: A DataFrame representing the individual builds of the
+        test jobs.
+    :returns: A DataFrame that has all of the original columns plus an
+        additional column named "duration_until_mergable" that contains the
+        amount of time from the start of __main_multijob until this sub-build
+        has been retried until success, assuming this was the final job in the
+        __main_multijob build.
+    """
+    grouped = build_data.groupby('job')
+    new_columns = None
+    for name, group in grouped:
+        last = None
+        group_rows = []
+        for row in group.sort_values(by='datetime',
```

Comment:
:question: Generally, manual loops in pandas code are not necessary. Unfortunately, I don't know pandas well enough to know how to make this look unnecessary.